### ISSUE TITLE
How to implement the Region-based loss?

### ISSUE THREAD

**Issue by QiushiYang** [MIC-DKFZ/nnUNet#458]
According to your latest paper: nnU-Net for Brain Tumor Segmentation for 2020-Brats Challenge, you use the Region-based training (loss) to improve the segmentation performance. I tried to implement it under the description of your paper (using  sigmoid BCE and dice losses for three final target regions), but got poor performance. I don't know why.
I cannot find the codes for it in this repo. Would you plan to share the implement codes of the region-based loss?

**Comment by FabianIsensee** [comment]
You need to convert the labels into overlapping regions before you pass them into the loss. 

The loss is here: https://github.com/MIC-DKFZ/nnUNet/blob/b26fbababfd79018f645026f5b5bf25dd42a11a8/nnunet/training/network_training/competitions_with_custom_Trainers/BraTS2020/nnUNetTrainerV2BraTSRegions.py#L73

The region conversion happens in the data augmentation, see here:

https://github.com/MIC-DKFZ/nnUNet/blob/b26fbababfd79018f645026f5b5bf25dd42a11a8/nnunet/training/network_training/competitions_with_custom_Trainers/BraTS2020/nnUNetTrainerV2BraTSRegions.py#L134

here: 
https://github.com/MIC-DKFZ/nnUNet/blob/b26fbababfd79018f645026f5b5bf25dd42a11a8/nnunet/training/data_augmentation/data_augmentation_moreDA.py#L136

and here:
https://github.com/MIC-DKFZ/nnUNet/blob/b26fbababfd79018f645026f5b5bf25dd42a11a8/nnunet/training/data_augmentation/data_augmentation_moreDA.py#L175

Just run the BraTS trainings and debug into them. Then it will become clear ;-)

**Comment by QiushiYang** [comment]
Thank you very much! I checked the codes for region-based loss I wrote before but didn't find wrongs, while the performance were worse after using it (either calculating 3 target tumor regions or further adding the background region). Could you please help me? The codes are showed as below. (I used the region-based loss consisting of dice and BCE losses both with sigmoid on the predicton maps in the channel-wise(class-wise) way.)

`class Loss_Region(torch.nn.Module):
    def __init__(self):
        super(Loss_Region, self).__init__()

    def forward(self, pred, target, use_dice=False, bce=True):
        loss = 0.0
        output_bg = pred[:,0,...]
        output_et = pred[:,3,...]
        output_tc = pred[:,1,...] + pred[:,3,...]
        output_wt = pred[:,1,...] + pred[:,2,...] + pred[:,3,...]

        b,h,w,d = target.shape
        target_region = target.clone().expand(4,b,h,w,d).clone()
        target_region = target_region.permute(1,0,2,3,4)
        
        target_region[:,0,...] = (target==0)
        target_region[:,1,...] = (target==4)
        target_region[:,2,...] = (target==1) + (target==4)
        target_region[:,3,...] = (target==1) + (target==2) + (target==4)

        if bce:
            _output_bg = F.sigmoid(output_bg)
            _output_et = F.sigmoid(output_et)
            _output_tc = F.sigmoid(output_tc) 
            _output_wt = F.sigmoid(output_wt)

            bce_loss_bg = bce_loss(_output_bg.clone(), target_region[:,0,...].cuda())
            bce_loss_et = bce_loss(_output_et.clone(), target_region[:,1,...].cuda())
            bce_loss_tc = bce_loss(_output_tc.clone(), target_region[:,2,...].cuda())
            bce_loss_wt = bce_loss(_output_wt.clone(), target_region[:,3,...].cuda())

            # _bce_loss = (bce_loss_bg + bce_loss_et + bce_loss_tc + bce_loss_wt )/4
            _bce_loss = (bce_loss_et + bce_loss_tc + bce_loss_wt )/3
            loss = loss + _bce_loss
            print ('Region_BCE loss: %f |'%_bce_loss.item(), end = " ")
        
        if use_dice:
            dice_loss_bg = dice(F.sigmoid(output_bg), target_region[:,0,...].cuda()==1)
            dice_loss_et = dice(F.sigmoid(output_et), target_region[:,1,...].cuda()==1)
            dice_loss_tc = dice(F.sigmoid(output_tc), target_region[:,2,...].cuda()==1)
            dice_loss_wt = dice(F.sigmoid(output_wt), target_region[:,3,...].cuda()==1)
            
            # _dice_loss = (dice_loss_bg + dice_loss_et + dice_loss_tc + dice_loss_wt)/4
            _dice_loss = (dice_loss_et + dice_loss_tc + dice_loss_wt)/3
            loss = loss + _dice_loss # criterions()
            print ('Region_Dice loss: %f |'%_dice_loss.item(), end = " ")
        
        del target_region
        torch.cuda.empty_cache()

        return loss`

`def bce_loss(prediction, label):
    label = label.clone().long()
    mask = label.clone().float()
    num_positive = torch.sum((mask==1).float()).float()
    num_negative = torch.sum((mask==0).float()).float()
    mask[mask == 1] = 1.0 * num_negative / (num_positive + num_negative)
    mask[mask == 0] = 1.0 * num_positive / (num_positive + num_negative)
    mask[mask == 2] = 0
    cost = torch.nn.functional.binary_cross_entropy(prediction.clone().float(),label.clone().float(), weight=mask, reduce=False)
    return torch.sum(cost) / (num_positive+1e-6)`

`def dice(output, target,eps =1e-5): # soft dice loss
    target = target.float()
    # num = 2*(output*target).sum() + eps
    num = 2*(output*target).sum()
    den = output.sum() + target.sum() + eps
    return 1.0 - num/den`

Thanks a lot!

**Comment by FabianIsensee** [comment]
Hi,
I do not have the time to go through your code and debug is. All I can do is refer you to my implementation which works as expected
Best,
Fabian
